### ISSUE TITLE
Fedora30,zfs-0.8.2-1 "cannot receive incremental stream: checksum mismatch or incomplete stream" on old snapshot

### ISSUE THREAD

**Issue by 4587658967** [openzfs/zfs#9555]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  Fedora
Distribution Version	|  30
Linux Kernel	|  5.3.7-200
Architecture	|  x86_64
ZFS Version	|  0.8.2-1
SPL Version	|  0.8.2-1

### Describe the problem you're observing
Hello. I am reluctant to open this issue since it may not be relevant to many users. This is also the first time I use GitHub so I will try to format my issue properly.

I encounter the following error when applying an incremental snapshot:
```
# zcat m/x\@2018-08-30_14-05-27.gz | ssh root@192.168.9.2 zfs receive -v m/x@2018-08-30_14-05-27
receiving incremental stream of m/x@2018-08-30_14-05-27 into m/x@2018-08-30_14-05-27
cannot receive incremental stream: checksum mismatch or incomplete stream
```
The error does not occur immediately, but after some time, somewhere within the stream. As the text suggests, the snapshot was created previously and saved to a file; unfortunately it does not exist in a pool anymore.

I was able to process the snapshot with ```zstreamdump```, which I understand to indicate that the snapshot is "valid" and that the error I received was incorrect.
```
# cat m-x@2018-08-30_14-05-27  | zstreamdump 
BEGIN record
        hdrtype = 1
        features = 4
        magic = 2f5bacbac
        creation_time = 5b88396a
        type = 2
        flags = 0x4
        toguid = c130e846178fbeb2
        fromguid = c5535ea78a7ee39d
        toname = m/x@2018-08-30_14-05-27
SUMMARY:
        Total DRR_BEGIN records = 1 (0 bytes)
        Total DRR_END records = 0 (0 bytes)
        Total DRR_OBJECT records = 12933 (2206568 bytes)
        Total DRR_FREEOBJECTS records = 1738 (0 bytes)
        Total DRR_WRITE records = 4376 (247398400 bytes)
        Total DRR_WRITE_BYREF records = 0 (0 bytes)
        Total DRR_WRITE_EMBEDDED records = 0 (0 bytes)
        Total DRR_FREE records = 13021 (0 bytes)
        Total DRR_SPILL records = 0 (0 bytes)
        Total records = 32069
        Total payload size = 249604968 (0xee0ab68)
        Total header overhead = 10005528 (0x98ac18)
        Total stream length = 259534208 (0xf782d80)
```
I have changed the name of the dataset in question to 'x'; I hope this is OK.

A number of such incremental snapshots had been applied already without issue. The datetime they were created is indicated as the snapshot name; in general there were several snapshots generated per day, and the last one is dated ```2018-08-29_23-00-03```. So in particular the snapshot at ```2018-08-29_23-00-03``` did not have any issues being applied. 
I note this just in case it might be an indication of a behavior which was introduced by a version change on the sender that may have occurred between those two times. If necessary, I may be able to establish whether such a version change took place, and specifically which version was in use, if the appropriate records still exist on my system(s).

I also tried applying the snapshot using zfs-fuse. In that case, the error message was different:
```
cannot receive incremental stream: invalid backup stream
```

Other notes:
- The system performing the receive is running inside a VM.
- The system which performed the send was Fedora, running zfs 0.7, likely 0.7.9 or there-abouts (_Edit: I am doubting whether this is correct, it may have actually been 0.7.12. I recall using 0.7.12 for some time._). I am not sure exactly but I may be able to determine this if I can locate the appropriate records.

```zdb``` on the receiving side:
```# zdb
m:
    version: 5000
    name: 'm'
    state: 0
    txg: 220405
    pool_guid: 15361807867333634321
    errata: 0
    hostname: 'localhost.localdomain'
    com.delphix:has_per_vdev_zaps
    vdev_children: 1
    vdev_tree:
        type: 'root'
        id: 0
        guid: 15361807867333634321
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 14971740363956603184
            path: '/dev/vdb1'
            whole_disk: 0
            metaslab_array: 30
            metaslab_shift: 33
            ashift: 12
            asize: 1099505598464
            is_log: 0
            DTL: 3263
            create_txg: 4
            com.delphix:vdev_zap_leaf: 45057
            com.delphix:vdev_zap_top: 45058
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data
```

```zdb``` on the sending side, **as it exists today** (not at 2018-08-30, when the snapshot was created)
```
m:
    version: 5000
    name: 'm'
    state: 0
    txg: 252131
    pool_guid: 8097840782044554886
    errata: 0
    hostname: 'localhost.localdomain'
    com.delphix:has_per_vdev_zaps
    vdev_children: 1
    vdev_tree:
        type: 'root'
        id: 0
        guid: 8097840782044554886
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 13979908019873797024
            path: '/dev/mapper/zfs'
            devid: 'dm-uuid-CRYPT-PLAIN-zfs'
            whole_disk: 0
            metaslab_array: 64
            metaslab_shift: 32
            ashift: 9
            asize: 439447191552
            is_log: 0
            create_txg: 4
            com.delphix:vdev_zap_leaf: 129
            com.delphix:vdev_zap_top: 130
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data
```

Since the application of the snapshot fails mid-file, I was thinking I could maybe isolate where the failure takes place:
```
# cat $SNAPFILE | tee --output-error=exit stream-out.dat | zfs receive -v $SNAP
```
and then compare the output of ```zstreamdump -v stream-out.dat``` and ```zstreamdump -v $SNAPFILE```, to indicate the block/location which would appear to trigger the error. I'm not sure if this would be useful. I am open to suggestions for whatever steps I should take.

### Describe how to reproduce the problem
Install Fedora 30, install zfsonlinux, apply the sequence of snapshots. Unfortunately I do not think I can share the snapshots as they are large and contain sensitive information. Aside from the installation of a few user-applications, the only actions I have taken on this newly installed system have been to install zfsonlinux.

### Include any warning/errors/backtraces from the system logs
I was not aware of any. I am available to provide any relevant information.

Thank you.

**Comment by GregorKopka** [comment]
Some general things:
Unless you want to change it: You don't have to specify the snapshot name on zfs recv, it'll default to the name in the stream - saves on typing.
Since you have the stream as a compressed file it could make more sense to send it as-is and decompress it on the target (less network traffic).

Regarding your problem:

Just to make sure you could zstreamump the prior snapshot file to make verify that the `toguid` there equals the `fromguid` in the one that fails. It's unlikely to be the problem (as you should get a different error message in case these differ) but shouldn't hurt to check.

To make sure that no network hickup causes this it could make sense to copy the file to the target (eg. using `rsync --inplace --no-whole-file --checksum` _twice_, the second run to safeguard against transmission errors) and afterwards retrying the recv with the local copy.

I'm a bit irritated by you doing ` zcat m/x\@2018-08-30_14-05-27.gz` on the failed transfer while doing a `cat m-x@2018-08-30_14-05-27` when doing the zstreamdump. Do you have both the compressed and an uncompressed version on the source? If so, try with the file zstreamdump sees as valid, instead of the compressed one?

**Comment by 4587658967** [comment]
Thank you. Sorry, I have an uncompressed version on the zfsonlinux host; the results are the same (for ```zfs recv``` and ```zstreamdump```) whether I use the compressed or uncompressed version. I was using them interchangeably in my post since it is the same underlying data (at least in theory). I did confirm that the local file has been transferred without errors.

And your comment about ```toguid``` and ```fromguid```, I have now confirmed that they match (the ```toguid``` from ```2018-08-29_23-00-03``` is the same as ```fromguid``` from ```2018-08-30_14-05-27```.
